File: temp_thkiengiangshoes/thkiengiangshoes-main/security/sanitizer.py
"""
TBS II — Sanitizer (XSS Prevention cho BI Dashboard & Response)
================================================================
Chong XSS cho du lieu hien thi len BI dashboard.
Mo rong validators.py co san — KHONG thay the.

Su dung:
  from security.sanitizer import (
      sanitize_html, sanitize_dashboard_data, sanitize_url,
      strip_dangerous_tags, SanitizerConfig
  )

  # Cho BI dashboard data:
  safe_data = sanitize_dashboard_data(raw_user_input)

  # Cho HTML response:
  safe_html = sanitize_html(user_content)
"""

import re
import html
from typing import Any, Dict, List, Optional, Union
from dataclasses import dataclass
from enum import Enum


class SanitizeMode(Enum):
    STRICT = "strict"      # Strip ALL HTML, chi giu text
    ALLOW_BASIC = "basic"  # Cho phep <b><i><u><p><br><ul><ol><li>
    DASHBOARD = "dashboard"  # Mode cho BI dashboard — triet de


@dataclass
class SanitizerConfig:
    mode: SanitizeMode = SanitizeMode.STRICT
    max_length: int = 10000
    strip_null_bytes: bool = True
    normalize_unicode: bool = True
    allow_urls: bool = False


# ============================================================
# XSS PATTERNS TO STRIP
# ============================================================

# Event handlers (onclick, onerror, onload, ...)
EVENT_HANDLER_PATTERN = re.compile(
    r'\bon\w+\s*=\s*["\'][^"\']*["\']|\bon\w+\s*=\s*[^\s>]+',
    re.IGNORECASE,
)

# Javascript protocol
JAVASCRIPT_PATTERN = re.compile(
    r'javascript\s*:', re.IGNORECASE
)

# Data URI voi script
DATA_URI_SCRIPT_PATTERN = re.compile(
    r'data\s*:\s*text\/html', re.IGNORECASE
)

# CSS expression (IE)
CSS_EXPRESSION_PATTERN = re.compile(
    r'expression\s*\(', re.IGNORECASE
)

# Script tags
SCRIPT_PATTERN = re.compile(
    r'<script[^>]*>.*?</script>', re.IGNORECASE | re.DOTALL
)

# Inline style voi url/javacript
INLINE_STYLE_DANGEROUS = re.compile(
    r'style\s*=\s*["\'][^"\']*(?:url|javascript|expression|behavior)[^"\']*["\']',
    re.IGNORECASE,
)

# Meta refresh redirect
META_REFRESH_PATTERN = re.compile(
    r'<meta[^>]*http-equiv\s*=\s*["\']?refresh["\']?[^>]*>',
    re.IGNORECASE,
)

# Object/embed tags
OBJECT_EMBED_PATTERN = re.compile(
    r'<(?:object|embed|applet|iframe|frame|frameset)[^>]*>.*?</(?:object|embed|applet|iframe|frame|frameset)>',
    re.IGNORECASE | re.DOTALL,
)

# Comment injection
COMMENT_INJECTION_PATTERN = re.compile(
    r'<!--.*?-->', re.DOTALL
)


# ============================================================
# SANITIZATION FUNCTIONS
# ============================================================

def _strip_null_bytes(text: str) -> str:
    """Xoa null bytes — tranh null byte injection."""
    return text.replace('\x00', '')


def _normalize_unicode(text: str) -> str:
    """Chuan hoa Unicode — tranh homoglyph attacks."""
    import unicodedata
    return unicodedata.normalize('NFKC', text)


def strip_dangerous_tags(html_text: str) -> str:
    """
    Xoa cac the HTML nguy hiem nhung giu lai noi dung text ben trong.
    """
    if not html_text:
        return ""

    # Remove script tags (ca noi dung)
    result = SCRIPT_PATTERN.sub('', html_text)

    # Remove object/embed/iframe (ca noi dung)
    result = OBJECT_EMBED_PATTERN.sub('', result)

    # Remove meta refresh
    result = META_REFRESH_PATTERN.sub('', result)

    # Remove event handlers
    result = EVENT_HANDLER_PATTERN.sub('', result)

    # Remove javascript: URLs
    result = JAVASCRIPT_PATTERN.sub('', result)

    # Remove CSS expressions
    result = CSS_EXPRESSION_PATTERN.sub('', result)

    # Remove inline styles with dangerous content
    result = INLINE_STYLE_DANGEROUS.sub('', result)

    # Remove data:text/html
    result = DATA_URI_SCRIPT_PATTERN.sub('', result)

    # Remove HTML comments
    result = COMMENT_INJECTION_PATTERN.sub('', result)

    return result.strip()


def sanitize_html(text: str, config: Optional[SanitizerConfig] = None) -> str:
    """
    Sanitize HTML input — mac dinh STRICT mode (strip tat ca HTML).
    """
    if not text:
        return ""

    cfg = config or SanitizerConfig()

    if cfg.strip_null_bytes:
        text = _strip_null_bytes(text)

    if cfg.normalize_unicode:
        text = _normalize_unicode(text)

    if cfg.mode == SanitizeMode.STRICT:
        # Strip EVERYTHING, chi giu text
        # Remove all HTML tags
        text = re.sub(r'<[^>]+>', '', text)
        # HTML entity decode + re-encode de sach
        text = html.escape(html.unescape(text))
        # Trim
        text = text.strip()

    elif cfg.mode == SanitizeMode.ALLOW_BASIC:
        # Strip dangerous tags first
        text = strip_dangerous_tags(text)
        # Cho phep basic formatting tags, strip con lai
        allowed_tags = ['b', 'i', 'u', 'p', 'br', 'ul', 'ol', 'li', 'strong', 'em', 'h3', 'h4']
        # Strip tat ca the khong duoc phep
        for tag in re.findall(r'</?(\w+)', text):
            if tag.lower() not in allowed_tags:
                text = re.sub(
                    rf'</?{re.escape(tag)}[^>]*>', '', text, flags=re.IGNORECASE
                )

    elif cfg.mode == SanitizeMode.DASHBOARD:
        # Aggressive sanitization cho BI dashboard
        text = strip_dangerous_tags(text)
        text = re.sub(r'<[^>]+>', '', text)  # Strip ALL HTML
        text = html.escape(text)
        text = text.strip()

    # Truncate if too long
    if len(text) > cfg.max_length:
        text = text[: cfg.max_length]

    return text


def sanitize_dashboard_data(data: Any) -> Any:
    """
    Sanitize du lieu truoc khi hien thi len BI dashboard.
    Xu ly ca string, dict, list, nested objects.
    """
    if isinstance(data, str):
        return sanitize_html(
            data,
            SanitizerConfig(mode=SanitizeMode.DASHBOARD, max_length=1000),
        )

    if isinstance(data, dict):
        return {k: sanitize_dashboard_data(v) for k, v in data.items()}

    if isinstance(data, list):
        return [sanitize_dashboard_data(item) for item in data]

    # Numbers, bool, None — pass through
    return data


def sanitize_url(url: str) -> str:
    """
    Validate va sanitize URL — chi cho phep http/https.
    """
    if not url:
        return ""

    url = url.strip()

    # Block javascript: protocol
    if re.match(r'^\s*javascript\s*:', url, re.IGNORECASE):
        return ""

    # Block data: protocol
    if re.match(r'^\s*data\s*:', url, re.IGNORECASE):
        return ""

    # Chi cho phep http/https
    if not url.lower().startswith(('http://', 'https://')):
        return ""

    # HTML encode URL
    url = html.escape(url, quote=True)

    if len(url) > 2048:
        return ""

    return url


def escape_json_for_html(data: Any) -> str:
    """
    Escape JSON data de nhung an toan vao HTML attribute.
    Tranh XSS qua JSON injection.
    """
    import json
    json_str = json.dumps(data, ensure_ascii=False)
    # Escape HTML entities
    json_str = json_str.replace('&', '\\u0026')
    json_str = json_str.replace('<', '\\u003c')
    json_str = json_str.replace('>', '\\u003e')
    json_str = json_str.replace("'", '\\u0027')
    return json_str


def safe_text_length(text: str, max_length: int = 5000) -> str:
    """
    Cat text an toan, khong cat giua chung ky tu Unicode.
    """
    if not text:
        return ""
    if len(text) <= max_length:
        return text
    # Tim diem cat an toan (space hoac newline)
    cut_point = max_length
    for i in range(max_length - 1, max(0, max_length - 100), -1):
        if text[i] in (' ', '\n', '\t', '.', '。', '、', ','):
            cut_point = i + 1
            break
    return text[:cut_point] + '…'

File: temp_thkiengiangshoes/thkiengiangshoes-main/security/sanitizer.ts
/**
 * TBS II — Sanitizer for Next.js / Cloudflare Workers
 * ====================================================
 * Chong XSS cho du lieu hien thi len BI dashboard.
 * Su dung: import { sanitizeDashboardData, sanitizeHtml } from '../security/sanitizer';
 */

type SanitizeMode = 'strict' | 'basic' | 'dashboard';

interface SanitizerConfig {
  mode: SanitizeMode;
  maxLength: number;
  stripNullBytes: boolean;
}

const DEFAULT_CONFIG: SanitizerConfig = {
  mode: 'strict',
  maxLength: 10000,
  stripNullBytes: true,
};

// XSS patterns
const SCRIPT_PATTERN = /<script[^>]*>[\s\S]*?<\/script>/gi;
const EVENT_HANDLER_PATTERN = /\bon\w+\s*=\s*(?:"[^"]*"|'[^']*'|[^\s>]+)/gi;
const JAVASCRIPT_URL_PATTERN = /javascript\s*:/gi;
const OBJECT_EMBED_PATTERN = /<(?:object|embed|applet|iframe|frame|frameset)[^>]*>[\s\S]*?<\/(?:object|embed|applet|iframe|frame|frameset)>/gi;
const META_REFRESH_PATTERN = /<meta[^>]*http-equiv\s*=\s*["']?refresh["']?[^>]*>/gi;
const DATA_URI_PATTERN = /data\s*:\s*text\/html/gi;

export function sanitizeHtml(text: string, config: Partial<SanitizerConfig> = {}): string {
  if (!text) return '';

  const cfg = { ...DEFAULT_CONFIG, ...config };

  let result = text;

  if (cfg.stripNullBytes) {
    result = result.replace(/\x00/g, '');
  }

  if (cfg.mode === 'strict' || cfg.mode === 'dashboard') {
    // Remove ALL dangerous constructs
    result = result.replace(SCRIPT_PATTERN, '');
    result = result.replace(OBJECT_EMBED_PATTERN, '');
    result = result.replace(META_REFRESH_PATTERN, '');
    result = result.replace(EVENT_HANDLER_PATTERN, '');
    result = result.replace(JAVASCRIPT_URL_PATTERN, '');
    result = result.replace(DATA_URI_PATTERN, '');

    // For dashboard: strip ALL tags
    if (cfg.mode === 'dashboard') {
      result = result.replace(/<[^>]+>/g, '');
    }

    // HTML entity encode
    result = escapeHtml(result);
  }

  if (result.length > cfg.maxLength) {
    result = result.substring(0, cfg.maxLength);
  }

  return result.trim();
}

function escapeHtml(text: string): string {
  const entityMap: Record<string, string> = {
    '&': '&amp;',
    '<': '&lt;',
    '>': '&gt;',
    '"': '&quot;',
    "'": '&#x27;',
    '/': '&#x2F;',
  };
  return text.replace(/[&<>"'/]/g, (ch) => entityMap[ch] || ch);
}

/**
 * Sanitize toan bo dashboard data (string, object, array).
 */
export function sanitizeDashboardData<T>(data: T): T {
  if (typeof data === 'string') {
    return sanitizeHtml(data, { mode: 'dashboard', maxLength: 1000 }) as unknown as T;
  }
  if (Array.isArray(data)) {
    return data.map(sanitizeDashboardData) as unknown as T;
  }
  if (data !== null && typeof data === 'object') {
    const sanitized: Record<string, unknown> = {};
    for (const [key, value] of Object.entries(data as Record<string, unknown>)) {
      sanitized[key] = sanitizeDashboardData(value);
    }
    return sanitized as unknown as T;
  }
  return data;
}

export function sanitizeUrl(url: string): string {
  if (!url) return '';
  const trimmed = url.trim();
  if (/^\s*javascript\s*:/i.test(trimmed)) return '';
  if (/^\s*data\s*:/i.test(trimmed)) return '';
  if (!trimmed.toLowerCase().startsWith('http://') && !trimmed.toLowerCase().startsWith('https://')) {
    return '';
  }
  if (trimmed.length > 2048) return '';
  return trimmed;
}

export function safeTextLength(text: string, maxLength: number = 5000): string {
  if (!text) return '';
  if (text.length <= maxLength) return text;
  let cutPoint = maxLength;
  for (let i = maxLength - 1; i > Math.max(0, maxLength - 100); i--) {
    if ([' ', '\n', '\t', '.', ',', '。', '、'].includes(text[i])) {
      cutPoint = i + 1;
      break;
    }
  }
  return text.substring(0, cutPoint) + '…';
}
